### PR TITLE
Implements cargo file locking using fcntl on Solaris.

### LINKED ISSUE #11421: cargo locking doesn't work on Solaris
### Problem

As described in https://bugzilla.mozilla.org/show_bug.cgi?id=1802095#c5 cargo locking doesn't work on Solaris:

"normally, running cargo twice in the same workspace will make one of them wait for the other".

Can you please point me to relevant peace of code? Maybe it's related to missing flock() system call on Solaris?!

### Steps

_No response_

### Possible Solution(s)

_No response_

### Notes

_No response_

### Version

_No response_

### PR DESCRIPTION
**Pull request by psumbera**
Fixes #11421.

### CODE REVIEW

File: src/cargo/util/flock.rs [hunk 1]
```diff
@@ -395,8 +395,38 @@ mod sys {
...
+        };
+        if (flag & libc::LOCK_SH) != 0 {
+            flock.l_type |= libc::F_RDLCK;
+        }
+        if (flag & libc::LOCK_EX) != 0 {
+            flock.l_type |= libc::F_WRLCK;
+        }
+        if (flag & libc::LOCK_UN) != 0 {
+            flock.l_type |= libc::F_UNLCK;
+        }
```
Comment (ehuss) [inline src/cargo/util/flock.rs:416]: I don't think l_type is a bitmask.  That is, F_RDLCK|F_WRLCK would result in 3 (F_UNLCK). Should this be `=` instead? That is, something like:

```rust

flock.l_type = if flag & libc::LOCK_UN != 0 { libc::F_RDLCK }
    else if flag & libc::LOCK_EX != 0 { libc::F_WRLCK }
    else if flag & libc::LOCK_SH != 0 { libc::F_RDLCK }
    else { panic!("…") };
```

Although here those flags won't get set simultaneously, I think it looks a little strange to use bitwise-or for something that is not a bitmask.

**Comment by ehuss** [comment]
Thanks! I don't have access to a Solaris system handy, so I can test this, but it seems like it should be correct to me.

@bors r+

File: src/cargo/util/flock.rs [hunk 2]
```diff
@@ -395,8 +395,38 @@ mod sys {
...
+            l_type: 0,
+            l_whence: 0,
+            l_start: 0,
+            l_len: 0,
+            l_sysid: 0,
+            l_pid: 0,
+            l_pad: [0, 0, 0, 0],
+        };
+        flock.l_type = if flag & libc::LOCK_UN != 0 {
+            libc::F_RDLCK
```
Comment (luqmana) [inline src/cargo/util/flock.rs:409]: Shouldn't this be `F_UNLCK`?

Comment (ehuss) [reply src/cargo/util/flock.rs:409]: Oh my, I didn't catch that. @psumbera can you follow up with a PR to update this?

Comment (psumbera) [reply src/cargo/util/flock.rs:409]: I will fix it. Sorry I didn't catch that. Since I have tested it this doesn't cause the problem with use in cargo.

Comment (psumbera) [reply src/cargo/util/flock.rs:409]: https://github.com/rust-lang/cargo/pull/11474

**Pull request merged**